Libraries/LibWeb/Editing/Internal/Algorithms.cpp

```diff
@@ -0,0 +1,1508 @@
+                if (fix_collapsed_space && collapse_spaces && offset_code_point == 0x20) {
+                    MUST(static_cast<DOM::CharacterData&>(*end_node).delete_data(end_offset, 1));
+                    continue;
+                }
+
+                // 2. Set collapse spaces to true if the end offsetth code unit of end node's data is a
+                //    space (0x0020), false otherwise.
+                collapse_spaces = offset_code_point == 0x20;
+
+                // 3. Add one to end offset.
+                ++end_offset;
+
+                // 4. Add one to length.
+                ++length;
+
+                // NOTE: We continue the loop here since we matched every condition from step 7.3
+                continue;
+            }
+        }
+
+        // 4. Otherwise, break from this loop.
+        break;
+    }
+
+    // 8. If fix collapsed space is true, then while (start node, start offset) is before (end node,
+    //    end offset):
+    if (fix_collapsed_space) {
+        while (true) {
+            auto relative_position = position_of_boundary_point_relative_to_other_boundary_point(*start_node, start_offset, *end_node, end_offset);
+            if (relative_position != DOM::RelativeBoundaryPointPosition::Before)
+                break;
+
+            // 1. If end node has a child in the same editing host with index end offset − 1, set end
+            //    node to that child, then set end offset to end node's length.
+            auto offset_minus_one_child = end_node->child_at_index(end_offset - 1);
+            if (offset_minus_one_child && is_in_same_editing_host(end_node, *offset_minus_one_child)) {
+                end_node = *offset_minus_one_child;
+                end_offset = end_node->length();
+                continue;
+            }
+
+            // 2. Otherwise, if end offset is zero and end node's parent is in the same editing host,
+            //    set end offset to end node's index, then set end node to its parent.
+            if (end_offset == 0 && is_in_same_editing_host(end_node, *end_node->parent())) {
+                end_offset = end_node->index();
+                end_node = *end_node->parent();
+                continue;
+            }
+
+            // 3. Otherwise, if end node is a Text node and its parent's resolved value for
+            //    "white-space" is neither "pre" nor "pre-wrap" and end offset is end node's length and
+            //    the last code unit of end node's data is a space (0x0020) and end node precedes a line
+            //    break:
+            auto* layout_node = end_node->parent()->layout_node();
+            if (layout_node && end_node->is_text() && end_offset == end_node->length() && precedes_a_line_break(end_node)) {
+                auto parent_white_space = layout_node->computed_values().white_space();
+
+                // FIXME: Find a way to get code points directly from the UTF-8 string
```

Comment:
The spec mentions code points, why would we then use UTF-16 code unit offsets?

Reply:
Last code point does also work (though, maybe it is slowly), but I mentioned that since it says "last code unit", which is just the last u16 in the array

Reply:
Which actually, when I think about it, does that need to be utf16 for this check? 🤔

Reply:
Ah, maybe not for this one, but perhaps for the general n'th code unit offset case

Reply:
It's reasons like this why I really want unicode-aware string-views, so you could just do `end_node_data.ends_with(' ');`. Since `0x20` is ASCII `' '` though, you could use `String::ends_with_bytes(" "sv);` in this case, or `ends_with(' ');`, but that one is `O(n)`. Then you won't have to `utf8_to_utf16`.

Reply:
Right - there is definitely some improvements we know we need to make in this space - but that's something that needs a bit of discussion with the maintainers on approach as there have been other ideas floated around on how to improve the inherent utf16-ness of the Web Platform and conversions back and forth we need to do with javascript.

I think the place I commented is a bit misleading, there are other cases in this file that we _do_ need `utf8_to_utf16` for the general case of getting it at the n'th offset.

Reply:
> there are other cases in this file that we do need utf8_to_utf16 for the general case of getting it at the n'th offset

We shouldn't have to. Our current `String` API requires it, yes, and it might indeed be more efficient to just cache the UTF-16 encodings in the future, but if you replace those `u32 {start,end}_offset`s with proper `UnicodeCodePointIterator`s then most of these operations would become `O(1)` rather than having to do an `O(n)` UTF-8 to UTF-16 conversion every time, and it certainly won't have to do any allocations.